### PR TITLE
[FLINK-5542] use YarnCluster vcores setting to do MaxVCore validation

### PR DESCRIPTION
**Pull request by leanken**
## What is the purpose of the change

See. [FLINK-5542](https://issues.apache.org/jira/browse/FLINK-5542)

use YarnCluster vcores setting to do MaxVCore validation instead of using the local yarn conf.

## Brief change log

- *Fetch MaxVCore num via yarnClient instead of using the local yarn conf*

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.
But I already reproduced the [Error/Exception](http://apache-flink-user-mailing-list-archive.2336050.n4.nabble.com/1-1-4-on-YARN-vcores-change-td11016.html), after the fix, it's working as expected.
	

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

**Comment by leanken** [comment]
cc @rehevkor5 @tillrohrmann

File: flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java [hunk 1]
```diff
@@ -282,18 +282,29 @@ private void isReadyForDeployment(ClusterSpecification clusterSpecification) thr
...
+		}
+
 		int configuredVcores = flinkConfiguration.getInteger(YarnConfigOptions.VCORES, clusterSpecification.getSlotsPerTaskManager());
 		// don't configure more than the maximum configured number of vcores
-		if (configuredVcores > numYarnVcores) {
+		if (configuredVcores > numYarnMaxVcores) {
 			throw new IllegalConfigurationException(
 				String.format("The number of virtual cores per node were configured with %d" +
-						" but Yarn only has %d virtual cores available. Please note that the number" +
+						" already exceeds the max vcores %d set on Yarn Cluster. Please note that the number" +
```
Comment (tillrohrmann) [inline flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java]: Maybe reword: "The number of requested virtual cores per node %d exceeds the maximum number virtual cores %d available in the Yarn cluster."

**Comment by leanken** [comment]
Thanks for your review. @tillrohrmann 
Resolved your comment. Looking forward for further work at FLINK community. About the proposal you mentioned on having a exhaustive check on job submission, I will try go through more submit scenarios and see if what we can do to achieve the exhaustive check goal and create a new follow up issue, thanks.

**Comment by GJL** [comment]
Hi, I am taking a look again.

File: flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java [hunk 2]
```diff
@@ -282,18 +282,29 @@ private void isReadyForDeployment(ClusterSpecification clusterSpecification) thr
 		}
 
 		// Check if we don't exceed YARN's maximum virtual cores.
-		// The number of cores can be configured in the config.
-		// If not configured, it is set to the number of task slots
-		int numYarnVcores = yarnConfiguration.getInt(YarnConfiguration.NM_VCORES, YarnConfiguration.DEFAULT_NM_VCORES);
+		// Fetch numYarnMaxVcores from all the RUNNING nodes via yarnClient
+		int numYarnMaxVcores = Integer.MIN_VALUE;
```
Comment (GJL) [inline flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java]: How about:
```
		final int numYarnMaxVcores;
		try {
			numYarnMaxVcores = yarnClient.getNodeReports(NodeState.RUNNING)
				.stream()
				.mapToInt(report -> report.getCapability().getVirtualCores())
				.max()
				.orElse(0);
		} catch (Exception e) {
			throw new YarnDeploymentException("Couldn't get cluster description, please check on the YarnConfiguration", e);
		}
```
?

Pros of doing so:
- `numYarnMaxVcores` is final
- If no NodeManagers are running at all, we would log `0` vcores instead of `Integer.MIN_VALUE`
- Concise & easy to read

File: flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java [hunk 3]
```diff
@@ -282,18 +282,29 @@ private void isReadyForDeployment(ClusterSpecification clusterSpecification) thr
...
-		if (configuredVcores > numYarnVcores) {
+		if (configuredVcores > numYarnMaxVcores) {
 			throw new IllegalConfigurationException(
-				String.format("The number of virtual cores per node were configured with %d" +
-						" but Yarn only has %d virtual cores available. Please note that the number" +
-						" of virtual cores is set to the number of task slots by default unless configured" +
-						" in the Flink config with '%s.'",
-					configuredVcores, numYarnVcores, YarnConfigOptions.VCORES.key()));
+				String.format("The number of requested virtual cores per node %d" +
+						" exceeds the maximum number virtual cores %d available in the Yarn Cluster." +
```
Comment (GJL) [inline flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java]: I think it  should be: [...] _maximum number **of** virtual cores_ [...]

File: flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java (hunk 2 above)
Comment (leanken) [reply flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java]: this way sure better than mine. I will do the update, thanks @GJL

File: flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java (hunk 3 above)
Comment (leanken) [reply flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java]: resolved exception msg typo

File: flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java (hunk 2 above)
Comment (leanken) [reply flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java]: update done and passed local verification

**Comment by leanken** [comment]
@GJL resolved yours comments.

**Comment by GJL** [comment]
LGTM, merging as soon as build is green.

**Comment by leanken** [comment]
@GJL Kindly remind. Build is ready to go.
